File: nova_lang/lib/nova/map.ex
defmodule Nova.Map do
  @moduledoc """
  Map operations for Nova runtime.
  Maps are represented as Elixir maps with keys being any term.
  """

  def empty, do: %{}

  def singleton(k, v), do: %{k => v}

  def insert(k, v, map), do: Map.put(map, k, v)

  def lookup(k, map) do
    case Map.fetch(map, k) do
      {:ok, v} -> {:just, v}
      :error -> :nothing
    end
  end

  def member(k, map), do: Map.has_key?(map, k)

  def delete(k, map), do: Map.delete(map, k)

  def keys(map), do: Map.keys(map)

  def values(map), do: Map.values(map)

  def union(m1, m2), do: Map.merge(m1, m2)

  def from_foldable(pairs) do
    Enum.reduce(pairs, %{}, fn {:tuple, k, v}, acc -> Map.put(acc, k, v) end)
  end

  def to_unfoldable(map) do
    Enum.map(map, fn {k, v} -> {:tuple, k, v} end)
  end

  def size(map), do: map_size(map)

  def is_empty(map), do: map_size(map) == 0
end

File: runtime/map.erl
-module(map).
-export([delete/2, empty/0, fromFoldable/1, insert/3, keys/1, lookup/2, 'map'/2, singleton/2, toUnfoldable/1, union/2, values/1]).

%% Empty map
empty() -> #{}.

%% Insert key-value pair
insert(K, V, M) -> maps:put(K, V, M).

%% Delete key
delete(K, M) -> maps:remove(K, M).

%% Get all keys
keys(M) -> maps:keys(M).

%% Lookup key (returns Maybe)
lookup(K, M) ->
    case maps:find(K, M) of
        {ok, V} -> {'_just', V};
        error -> '_nothing'
    end.

%% Create singleton map
singleton(K, V) -> #{K => V}.

%% Union of two maps (right-biased)
union(M1, M2) -> maps:merge(M1, M2).

%% Get all values
values(M) -> maps:values(M).

%% Create map from list of tuples [{_tuple, K, V}]
fromFoldable(List) ->
    lists:foldl(fun({'_tuple', K, V}, Acc) -> maps:put(K, V, Acc) end, #{}, List).

%% Convert map to list of tuples [{_tuple, K, V}]
toUnfoldable(M) ->
    [{'_tuple', K, V} || {K, V} <- maps:to_list(M)].

%% Map a function over values in a map (Functor instance for Map)
'map'(F, M) -> maps:map(fun(_K, V) -> F(V) end, M).
